solr/core/src/test/org/apache/solr/update/TestNestedUpdateProcessor.java

```diff
@@ -0,0 +1,171 @@
+      "            ]\n" +
+      "        }\n" +
+      "    }\n" +
+      "}";
+
+  private static final String errDoc = "{\n" +
+      "    \"add\": {\n" +
+      "        \"doc\": {\n" +
+      "            \"id\": \"1\",\n" +
+      "            \"children" + PATH_SEP_CHAR + "a\": [\n" +
+      "                {\n" +
+      "                    \"id\": \"2\",\n" +
+      "                    \"foo_s\": \"Yaz\"\n" +
+      "                    \"grandChild\": \n" +
+      "                          {\n" +
+      "                             \"id\": \""+ grandChildId + "\",\n" +
+      "                             \"foo_s\": \"Jazz\"\n" +
+      "                          },\n" +
+      "                },\n" +
+      "                {\n" +
+      "                    \"id\": \"3\",\n" +
+      "                    \"foo_s\": \"Bar\"\n" +
+      "                }\n" +
+      "            ]\n" +
+      "        }\n" +
+      "    }\n" +
+      "}";
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    initCore("solrconfig-update-processor-chains.xml", "schema15.xml");
+  }
+
+  @Before
+  public void before() throws Exception {
+    assertU(delQ("*:*"));
+    assertU(commit());
+  }
+
+  @Test
+  public void testDeeplyNestedURPGrandChild() throws Exception {
+    indexSampleData(jDoc);
+
+    assertJQ(req("q", IndexSchema.PATH_FIELD_NAME + ":*" + PATH_SEP_CHAR + "grandChild",
+        "fl","*",
+        "sort","id desc",
+        "wt","json"),
+        "/response/docs/[0]/id=='" + grandChildId + "'");
+  }
+
+  @Test
+  public void testDeeplyNestedURPChildren() throws Exception {
+    final String[] childrenTests = {"/response/docs/[0]/id=='" + childrenIds[0] + "'", "/response/docs/[1]/id=='" + childrenIds[1] + "'"};
+    indexSampleData(jDoc);
+
+    assertJQ(req("q", IndexSchema.PATH_FIELD_NAME + ":children",
+        "fl","*",
+        "sort","id asc",
+        "wt","json"),
+        childrenTests);
+  }
+
+  @Test
+  public void testDeeplyNestedURPChildrenWoId() throws Exception {
+    final String parentId = "3";
+    final String childKey = "grandChild";
+    final String expectedId = parentId + PATH_SEP_CHAR + childKey + PATH_SEP_CHAR + "0";
+    SolrInputDocument noIdChildren = sdoc("id", "1", "children", sdocs(sdoc("id", "2", "name_s", "Yaz"), sdoc("id", parentId, "name_s", "Jazz", childKey, sdoc("names_s", "Gaz"))));
+    UpdateRequestProcessor nestedUpdate = new NestedUpdateProcessorFactory().getInstance(req(), null, null);
+    AddUpdateCommand cmd = new AddUpdateCommand(req());
+    cmd.solrDoc = noIdChildren;
+    nestedUpdate.processAdd(cmd);
+    cmd.clear();
+    List children = (List) noIdChildren.get("children").getValues();
+    SolrInputDocument idLessChild = (SolrInputDocument)((SolrInputDocument) children.get(1)).get(childKey).getValue();
+    assertTrue("Id less child did not get an Id", idLessChild.containsKey("id"));
```

Comment:
This is fine but FYI doing a doc.toString() and doing a simple assertEquals on that String is often more thorough and easy to read and less code.  That said, sometimes it can be brittle to unrelated changes.

Reply:
come to think of it, in this whole test file I don't see a test for what the `_NEST_PATH_` field is supposed to look like?  That's kinda important!  If you modify this to test via doc.toString it'd cover it.  Granted `testDeeplyNestedURPChildren()` or some new method that ought to test this as well.

Reply:
I have just added a new commit with tests to ensure this :)